strings/suffix_automaton.py

```diff
@@ -0,0 +1,150 @@
+from __future__ import annotations
+
+
+class SuffixAutomaton:
+    """
+    Suffix Automaton data structure for efficient substring counting and searching.
+
+    Usage:
+    sa = SuffixAutomaton()
+    sa.extend('abc')
+    terminal_states = sa.terminal()
+    is_found = sa.find('ab')
+
+    :param alpha: Number of characters in the alphabet
+    (default is 26 for lowercase English letters)
+    :param first: ASCII value of the first character in the alphabet (default is 'a')
+    """
+
+    def __init__(self, alpha: int = 26, first: int = ord("a")):
+        self.alpha = alpha
+        self.first = first
+        self.substrings = 0
+        self.last = 0
+        self.st = [self.Node(0, -1, [0] * self.alpha)]
+
+    class Node:
+        def __init__(self, length: int, link: int, next_states: list[int]):
+            self.len = length
+            self.link = link
+            self.next = next_states
+
+    def extend(self, c: str) -> None:
```

Comment:
Is it possible to add a PR and Change its doctests at the same, guess not only 1 at a time!!

Reply:
If we are adding a new file, we should add tests in the same PR.
If we are modifying an existing file, we should modify the code OR the tests but not both in the same PR.